pamelalozano/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -96,6 +93,52 @@
       return eventsMap;
   }
 
+  /*
+  * This function will merge all the occupied times of the mandatory attendees
+  * into a single list of TimeRanges, that way I can create a lit of available 
+  * times that they all have in common.
+  *
+  * 
+  * Runtime: O(p * m) where p is the number of attendees and m the number of events
+  * 
+  * @param mandatoryAttendees   A list of the mandatory attendes
+  * @param eventsMap            A map with the keys as the name of the attendees 
+  *                             and the values as a list with the occupied time ranges
+  *
+  * returns  a list of the available times of the request
+  */
+  public List<TimeRange> checkAvailability(HashSet<String> mandatoryAttendees, HashMap<String, List<TimeRange>> eventsMap){
+    
+    LinkedList<TimeRange> timeOccupied = new LinkedList<>();
+    for (String person : mandatoryAttendees) {
+        List<TimeRange> schedule = eventsMap.get(person);
+        for(TimeRange range : schedule){
+          timeOccupied.add(range);
+       }
+    }
+    //ToDo: Merge times
+
+    List<TimeRange> timeAvailable = new ArrayList<>();
+    TimeRange currentEvent = null;
+    for (ListIterator<TimeRange> listIterator = timeOccupied.listIterator(); ;) {
+        if(!listIterator.hasPrevious()) {
+            currentEvent = listIterator.next();
+            timeAvailable.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.start(), false));
+        }
+
+        if(!listIterator.hasNext()) {
+            timeAvailable.add(TimeRange.fromStartEnd(currentEvent.end(), TimeRange.END_OF_DAY, true));
+            break;
+        } else {
+             int endOfCurrentEvent = currentEvent.end();
+             currentEvent = listIterator.next();
+             timeAvailable.add(TimeRange.fromStartEnd(endOfCurrentEvent, currentEvent.start(), false));
```

Comment:
Can the next event start before `endOfCurrentEvent`? It looks like this would add a TimeRange with negative duration in that case. Also I think this assumes the `timeOccupied` events are sorted. Consider this input: `[Range[7, 10), Range[2, 4)]`

Reply:
+1 about sorting. Assuming that you have merged all of the available times, you shouldn't have events that overlap at all, right?